modmail/dispatcher.py

```diff
@@ -0,0 +1,167 @@
+import asyncio
+import bisect
+import logging
+from typing import Callable, Coroutine, Dict, List, Optional, Union
+
+from modmail import ModmailLogger
+
+logger: ModmailLogger = logging.getLogger(__name__)
+
+CoroutineFunction = Callable[..., Coroutine]
+
+
+class Dispatcher:
+    """
+    Dispatches events through an async event handler system.
+
+    Supports blocking events and priority.
+    """
+
+    # These are separate because it makes using the bisect module easier. See _register_handler.
+    blocking_handlers: Dict[str, List[Optional[CoroutineFunction]]]
+    blocking_priorities: Dict[str, List[int]]
+    handlers: Dict[str, List[CoroutineFunction]]
+
+    def __init__(self, *event_names: str):
+        self.handlers = {}
+        self.blocking_handlers = {}
+        self.blocking_priorities = {}
+
+        self.register_event(*event_names)
+
+    def register_event(self, *event_names: str) -> None:
```

Comment:
Perhaps this could be more clear, as this can take multiple events, but the name of the function is singular.

Reply:
👍